### PR TITLE
Add guard for non-default stream in DDP's autograd engine callback

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40115 Add guard for non-default stream in DDP's autograd engine callback**

Closes https://github.com/pytorch/pytorch/issues/37790
Closes https://github.com/pytorch/pytorch/issues/37944

A user may wish to run DDP's forward + backwards step under a non-default CUDA stream such as those created by `with torch.cuda.Stream(stream)`. In this case, the user should be responsible for synchronizing events on this stream with other streams used in the program (per the documentation at https://pytorch.org/docs/stable/notes/cuda.html#cuda-semantics), but currently DDP has a bug which causes DDP under non-default streams to fail.

If a user does the following:
```
stream = torch.cuda.Stream()
with torch.cuda.stream(stream):
model = DDP(...)
loss = model(inptut).sum()
loss.backward()
grad = model.module.weight.grad()
average = dist.all_reduce(grad)
```

There is a chance that `average` and `grad` will not be equal. This is because the CUDA kernels corresponding to the  `all_reduce` call may run before `loss.backward()`'s kernels are finished. Specifically, in DDP we copy the allreduced gradients back to the model parameter gradients in an autograd engine callback, but this callback runs on the default stream. Note that this can also be fixed by the application synchronizing on the current stream, although this should not be expected, since the application is not using the current stream at all.

This PR fixes the issue by passing the current stream into DDP's callback.

Tested by adding a UT `test_DistributedDataParallel_non_default_stream` that fails without this PR

Differential Revision: [D22073353](https://our.internmc.facebook.com/intern/diff/D22073353/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22073353/)!

### CODE REVIEW

File: torch/csrc/distributed/c10d/reducer.cpp [hunk 1]
```diff
@@ -400,8 +401,17 @@ void Reducer::mark_variable_ready(VariableIndex index) {
     }
     local_used_work_ = process_group_->allreduce(local_used_maps_dev_);
 
+    // The autograd engine uses the default stream when running callbacks, so we
+    // pass in the current CUDA stream in case it is not the default.
+    c10::DeviceType deviceType = replica.contents.device().type();
+    const c10::impl::VirtualGuardImpl guard =
+        c10::impl::VirtualGuardImpl{deviceType};
+    const c10::Stream currentStream =
+        guard.getStream(replica.contents.device());
```
Comment (mrshenli) [inline torch/csrc/distributed/c10d/reducer.cpp:410]: Does this still work if this is CPU only DDP? Or is it because we need to support DDP so that we cannot directly call `getCurrentCUDAStream` from `CUDAStream`?

Comment (rohan-varma) [reply torch/csrc/distributed/c10d/reducer.cpp:410]: Yes, we could've done this with `at::cuda::getCurrentCudaStream` and then set that as the guard below, although this wouldn't work for the CPU use case.

This approach here works for CPU or GPU, as the `VirtualGuardImpl` simply dispatches the call to the relevant guard implementation `CPUGuardImpl` for CPU and `CUDAGuardImpl` for CUDA. The former is effectively a no-op since there isn't a concept of streams in CPU.

CPU DDP should work fine with this, and I verified by running the DDP CPU test as well.

**Review by mrshenli — APPROVED** [review]
Thanks for fixing!
